### ISSUE TITLE
bug in ble2mqtt_task() function

### ISSUE THREAD

**Issue by HkiriSiwar** [shmuelzon/esp32-ble2mqtt#194]
Hello, 
I have a decrease in heap memory storage in real time. When I check this problem, I see that the ble2mqtt_task() function is responsible for this problem. Can you check where the bug is in this function?

**Comment by shmuelzon** [comment]
Hey,

Can you please provide a bit more information? I have these devices running for weeks without any memory leak.
How did you reach the conclusion that this function is to blame? Almost everything goes through it...
Do you have anything special in your setup? Many BLE devices Broadcasters? Anything?
Are you limiting the number of devices you connect to?
During normal operation, we allocate memory for every detected BLE device so, if there are many, we may run out. There's a timer in the background to clean up stale devices but maybe it's not enough in your setup.

This project publishes, in the MQTT bus the amount of free memory every minute. Any chance you can plot that in a graph and share it here?

Thanks

**Comment by dakhnod** [comment]
Heyo, I recorded and got these results:

https://docs.google.com/spreadsheets/d/19c_Fgn2I4E9Yvk6uMY0I3Hcvs3AkzbvZ8puYxrG_Ebg/edit?usp=sharing

Does the sawtooth stem from the discovery of BLE devices?

**Comment by shmuelzon** [comment]
Generally speaking, yes. Each discovered BLE device creates an entry for it in memory. There's a "background task" that periodically cleans up the list from devices that we're not connected to. Looking at your graph, I don't see a downward trend or something that might point to a memory leak

**Comment by dakhnod** [comment]
@shmuelzon thanks for coming back to me. There really is no trend, but I would assess that the graph comes dangerously close to 0...

**Comment by shmuelzon** [comment]
Agreed, it depends on how "noisy" your environment is. For a quick work-around, you can have the task run more often, just change the timer here: https://github.com/shmuelzon/esp32-ble2mqtt/blob/e66ec5aad03991f2557be93d2e625b71dfdd9289/main/ble.c#L1044

Otherwise, we'd need some other mechanism to poll the free memory more often and purge the list if we reach some critical limit.

**Comment by dakhnod** [comment]
I am only using the bridge for BLE connections. As such, an option to disable the discovery altogether would be nice.

Apart from that, can we somehow just store the hash of a device in stead of the whole information?
Maybe I am misunderstanding the reason for storing advertisements wrong though...

**Comment by shmuelzon** [comment]
It's been a while but, IIRC, the BLE layer/file is meant to be generic so it doesn't care what the application `ble2mqtt.c` wants to do with the devices, it just notifies it for events and waits for actions/commands for it to do. We keep a list of all discovered devices so we can notify the application of each discovered device only once. Otherwise we'd just send it an even every few milliseconds. The size of the structure isn't exactly small at ~48 bytes but... 🤷‍♂️

I don't think we can disable discovery altogether since, otherwise, we won't detect the devices we do want to connect to.

**Comment by dakhnod** [comment]
I agree, 48 bytes should be no problem. Reducing this to some CRC16 could help, but doesn't feel like a proper fix.

You are right that we can not disable the general BLE discovery.
But maybe we can disable the consumer that does the `malloc` and MQTT publishes?

**Comment by dakhnod** [comment]
From a quick look I would assume that commenting out this line should suffice?

https://github.com/shmuelzon/esp32-ble2mqtt/blob/master/main/ble2mqtt.c#L967

**Comment by shmuelzon** [comment]
That'll disable MQTT messages for broadcaster devices, i.e., those devices that we don't connect to but just publish whatever they broadcast (ibeacons, sensors, etc.) but I don't think that'll help with the memory usage

**Comment by dakhnod** [comment]
Hey, you are right, the memory still shows a sawtooth.
What would be the way to disable the mallocs?

**Comment by shmuelzon** [comment]
Do you need to be able to connect to devices or just events from broadcasters?

**Comment by dakhnod** [comment]
I need to be able to connect, exclusively

**Comment by dakhnod** [comment]
Also, for the random reboots, I set up debug logging, hopefully will be able to debug there

**Comment by shmuelzon** [comment]
Well, in order to connect to a discovered device, we need to save its information (with the current design, as discovery is separate from connection).

If you want a workaround, you can add something right before https://github.com/shmuelzon/esp32-ble2mqtt/blob/e66ec5aad03991f2557be93d2e625b71dfdd9289/main/ble.c#L754
```c
if (!config_ble_should_connect(mactoa(param->scan_rst.bda)))
    break;
```
(You'll probably also need to include `config.h`)
With that it should only save devices which we'll connected to

**Comment by dakhnod** [comment]
I see. I assume we need to save the discovered device in order to pass it from the "discovery layer" to the "connection layer"?

**Comment by shmuelzon** [comment]
Sort of, yes. Everything is asynchronous because we don't want to block any thread. Also, for connecting to a device, we need a bit more information than just the MAC address so that needs to be saved somewhere. The `ble.c` code was meant to act as an abstraction layer to the main application and hide the small details from it, so that also meant saving some BLE-related information from previous scans/detections.
